src/main/java/org/inferred/freebuilder/processor/excerpt/CheckedBiMap.java

```diff
@@ -0,0 +1,237 @@
+package org.inferred.freebuilder.processor.excerpt;
+
+import com.google.common.collect.BiMap;
+import org.inferred.freebuilder.processor.source.Excerpt;
+import org.inferred.freebuilder.processor.source.LazyName;
+import org.inferred.freebuilder.processor.source.SourceBuilder;
+import org.inferred.freebuilder.processor.source.ValueType;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+
+public class CheckedBiMap extends ValueType implements Excerpt {
+
+  public static final LazyName TYPE = LazyName.of("CheckedBiMap", new CheckedBiMap());
+
+  private CheckedBiMap() {
+  }
+
+  @Override
+  public void addTo(SourceBuilder code) {
+    code.addLine("")
+        .addLine("/**")
+        .addLine(" * A bimap implementation that delegates to a provided put method")
+        .addLine(" * to perform entry validation and insertion into a backing bimap.")
+        .addLine(" */")
+        .addLine("private static class %s<K, V> extends %s<K, V> implements %s<K, V> {",
+            TYPE, AbstractMap.class, BiMap.class
+        )
+        .addLine("")
+        .addLine("  private final %s<K, V> bimap;", BiMap.class)
+        .addLine("  private final %s<K, V> put;", BiConsumer.class)
+        .addLine("")
+        .addLine("  %s(%s<K, V> bimap, %s<K, V> put) {", TYPE, BiMap.class, BiConsumer.class)
+        .addLine("    this.bimap = bimap;")
+        .addLine("    this.put = put;")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V get(Object key) {")
+        .addLine("    return bimap.get(key);")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public boolean containsKey(Object key) {")
+        .addLine("    return bimap.containsKey(key);")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V put(K key, V value) {")
+        .addLine("    V oldValue = bimap.get(key);")
+        .addLine("    put.accept(key, value);")
+        .addLine("    return oldValue;")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V forcePut(K key, V value) {")
+        .addLine("    V oldValue = bimap.get(key);")
+        .addLine("    put.accept(key, value);")
+        .addLine("    return oldValue;")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V remove(Object key) {")
+        .addLine("    return bimap.remove(key);")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public void clear() {")
+        .addLine("    bimap.clear();")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public %s<%s<K, V>> entrySet() {",
+            Set.class, Map.Entry.class
+        )
+        .addLine("    return new %s<>(bimap.entrySet(), put);", CheckedBiMap.CheckedEntrySet.TYPE)
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public %s<V,K> inverse() {", BiMap.class)
+        .addLine("    return bimap.inverse();")
```

Comment:
This is returning a mutable view, which bypasses the checks. You'll need to return another CheckedBiMap with a reversed `put` instance. Also, you'll need tests that the checks are performed successfully via this route.